apis/python/src/tiledbsoma/io/_caching_reader.py

```diff
@@ -0,0 +1,199 @@
+            ]
+            for block_idx in missing_blocks:
+                self._cache_stats[block_idx].miss += 1
+                self._cache[block_idx] = self._read_block(block_idx)
+
+            requested_blocks = [
+                self._cache[block_idx] for block_idx in range(start_block, end_block)
+            ]
+
+            self._mark_and_sweep_blocks(start_block, end_block)
+
+        return requested_blocks
+
+    def _mark_and_sweep_blocks(self, start: int, stop: int) -> None:
+        for block_idx in range(start, stop):
+            self._cache.move_to_end(block_idx)
+            self._cache_stats[block_idx].hit += 1
+
+        for i in range(max(0, len(self._cache) - self._max_cache_blocks)):
+            self._cache.popitem(last=False)
+
+    def _reset_cache(self) -> None:
+        with self._cache_lock:
+            self._cache.clear()
+
+    def read(self, size: int = -1) -> bytes:
+        if size is None:
+            size = -1  # type: ignore[unreachable]
+        if size < 0:
+            size = self._file_length - self._pos
+        if size == 0:
+            return b""
+
+        blocks = self._load_cache(self._pos, self._pos + size)
+
+        start = self._pos % self._cache_block_size
+        end = start + size
+        arr = pa.chunked_array(blocks)[start:end].combine_chunks()  # NB: copy
+        assert arr.offset == 0
+        b = arr.buffers()[1].to_pybytes()  # NB: copy
+        self._pos += len(b)
+        return cast(bytes, b)
+
+    def readinto(self, buf: WritableBuffer) -> int | None:
+        if not isinstance(buf, memoryview):
+            buf = memoryview(buf)
+        if buf.nbytes == 0:
+            return 0
+        buf = buf.cast("B")
+
+        size = buf.nbytes
+        blocks = self._load_cache(self._pos, self._pos + size)
+        start = self._pos % self._cache_block_size
+        end = start + size
+        carr = pa.chunked_array(blocks)[start:end]
+
+        dst = ctypes.addressof(ctypes.c_char.from_buffer(buf))
+        dstidx = 0
+        for c in carr.chunks:
+            src = c.buffers()[1].address + c.offset
+            count = len(c)
+            ctypes.memmove(dst + dstidx, src, count)
+            dstidx += count
+
+        assert dstidx == len(carr)
+        self._pos += dstidx
+        return dstidx
+
+    def close(self) -> None:
+        self._reset_cache()
+        self._file.close()
+
+    def tell(self) -> int:
+        return self._pos
+
+    def seek(self, offset: int, whence: int = 0) -> int:
+        if whence not in [io.SEEK_SET, io.SEEK_CUR, io.SEEK_END]:
+            raise ValueError("Invalid whence value {whence})")
+
+        if whence == io.SEEK_END:
+            offset = self._file_length + offset
+        elif whence == io.SEEK_CUR:
+            offset += self._pos
+        elif whence == io.SEEK_SET:
+            offset = offset
+
+        if offset < 0:
+            raise OSError("seek() returned invalid position")
```

Comment:
This doesn't match the `IOBase` definition of `seek` for `io.SEEK_END` and `io.SEEK_CUR`:

> Change the stream position to the given byte offset, interpreted relative to the position indicated by whence, and return the new absolute position. Values for whence are:

    [os.SEEK_SET](https://docs.python.org/3/library/os.html#os.SEEK_SET) or 0 – start of the stream (the default); offset should be zero or positive

    [os.SEEK_CUR](https://docs.python.org/3/library/os.html#os.SEEK_CUR) or 1 – current stream position; offset may be negative

    [os.SEEK_END](https://docs.python.org/3/library/os.html#os.SEEK_END) or 2 – end of the stream; offset is usually negative

Reply:
which part doesn't match?  it looks like they match to my eye, but I'm probably missing something

Reply:
I'll add, whatever mistake I'm making must have been replicated into the test code, as I explicitly tested it against a Python built-in file and got the same results.   Weird....

Reply:
I have slightly refactored the code for clarity - it is exactly the same logic, but (I believe) easier to read.  Let me know if you still see an issue

Reply:
I was wrong about there being a mismatch. Both versions look fine, I just flipped the offset and position in my head.

Reply:
lol, I've been there.  The good news is that it is now easier to read, so an improvement I think.